### ISSUE TITLE
Assert error in vbf_stp_condfetch(), cache/cache_fetch.c line 775

### ISSUE THREAD

**Issue by BastienBalaud** [varnishcache/varnish-cache#3614]
Hello,

Recently one of our backend was on heavy load(DDOS) and we had 2 panic on ours Varnish server(two different servers encounter the panic with 8 minutes of interval). We running Varnish 6.5.1 on CentOS 8 (8.3.1-5). For the sotrage we use malloc. The two crashes are not caused by the same request (but they targeted the same backend). I replay both request without encounter any crash. 

I have attached the full log with the issue.
Panic 1 : 
```
varrnishd[1754]: Child (4464) not responding to CLI, killed it.
varrnishd[1754]: Unexpected reply from ping: 400 CLI communication error (hdr)
varrnishd[1754]: Child (4464) not responding to CLI, killed it.
varrnishd[1754]: Unexpected reply from ping: 400 CLI communication error
varrnishd[1754]: Child (4464) died signal=6 (core dumped)
varrnishd[1754]: Child (4464) Panic at: Wed, 12 May 2021 10:14:38 GMT
Assert error in vbf_stp_condfetch(), cache/cache_fetch.c line 775:
  Condition((stale_oc->flags & OC_F_FAILED) == 0) not true.
version = varnish-6.5.1 revision 1dae23376bb5ea7a6b8e9e4b9ed95cdc9469fb64, vrt api = 12.0
ident = Linux,4.18.0-240.22.1.el8_3.x86_64,x86_64,-junix,-sdefault,-sdefault,-hcritbit,epoll
now = 164579.801494 (mono), 1620814474.232021 (real)
Backtrace:
  0x562cf7b39d9c: /usr/sbin/varnishd(+0x52d9c) [0x562cf7b39d9c]
  0x562cf7bad33b: /usr/sbin/varnishd(VAS_Fail+0x1b) [0x562cf7bad33b]
  0x562cf7b26d44: /usr/sbin/varnishd(+0x3fd44) [0x562cf7b26d44]
  0x562cf7b5ef48: /usr/sbin/varnishd(+0x77f48) [0x562cf7b5ef48]
  0x562cf7b5f458: /usr/sbin/varnishd(+0x78458) [0x562cf7b5f458]
  0x7f369823714a: /lib64/libpthread.so.0(+0x814a) [0x7f369823714a]
  0x7f3697f68f23: /lib64/libc.so.6(clone+0x43) [0x7f3697f68f23]
pthread.self = 0x7f368de68700
pthread.name = (cache-worker)
pthread.attr = {
  guard = 4096,
  stack_bottom = 0x7f368de5b000,
  stack_top = 0x7f368de69000,
  stack_size = 57344,
}
thr.req = (nil) {
},
thr.busyobj = 0x7f34f8e73660 {
  end = 0x7f34f8e83640,
  retries = 0,
  sp = 0x7f3637478d20 {
    fd = -1, vxid = 63863376,
    t_open = 1620814473.176161,
    t_idle = 1620814473.176344,
    ws = 0x7f3637478d60 {
      id = \"ses\",
      {s, f, r, e} = {0x7f3637478db8, +96, (nil), +576},
    },
    transport = HTTP/1 {
      state = HTTP1::NewReq
    }
    client = 127.0.0.1 26444 127.0.0.1:6081,
  },
  worker = 0x7f368de66570 {
    ws = 0x7f368de66618 {
      id = \"wrk\",
      {s, f, r, e} = {0x7f368de65b60, +0, (nil), +2040},
    },
    VCL::method = BACKEND_RESPONSE,
    VCL::return = deliver,
    VCL::methods = {BACKEND_FETCH, BACKEND_RESPONSE},
  },
  vfc = 0x7f34f8e75568 {
    failed = 0,
    req = 0x7f34f8e73c78,
    resp = 0x7f34f8e740f0,
    wrk = 0x7f368de66570,
    oc = 0x7f364a882200,
    obj_flags = 0x0,
  },
  ws = 0x7f34f8e736a0 {
    id = \"bo\",
    {s, f, r, e} = {0x7f34f8e755b0, +3720, (nil), +57480},
  },
  ws_bo = 0x7f34f8e75b18,
  http[bereq] = 0x7f34f8e73c78 {
    ws = 0x7f34f8e736a0 {
      [Already dumped, see above]
    },
... 
```
Panic 2
```
varnishd[1575]: Child (3200) not responding to CLI, killed it.
varnishd[1575]: Unexpected reply from ping: 400 CLI communication error (hdr)
varnishd[1575]: Child (3200) not responding to CLI, killed it.
varnishd[1575]: Unexpected reply from ping: 400 CLI communication error
varnishd[1575]: Child (3200) died signal=6 (core dumped)
varnishd[1575]: Child (3200) Panic at: Wed, 12 May 2021 10:22:43 GMT
Assert error in vbf_stp_condfetch(), cache/cache_fetch.c line 775:
  Condition((stale_oc->flags & OC_F_FAILED) == 0) not true.
version = varnish-6.5.1 revision 1dae23376bb5ea7a6b8e9e4b9ed95cdc9469fb64, vrt api = 12.0
ident = Linux,4.18.0-240.22.1.el8_3.x86_64,x86_64,-junix,-sdefault,-sdefault,-hcritbit,epoll
now = 164725.388570 (mono), 1620814958.928224 (real)
Backtrace:
  0x55f51c0f8d9c: /usr/sbin/varnishd(+0x52d9c) [0x55f51c0f8d9c]
  0x55f51c16c33b: /usr/sbin/varnishd(VAS_Fail+0x1b) [0x55f51c16c33b]
  0x55f51c0e5d44: /usr/sbin/varnishd(+0x3fd44) [0x55f51c0e5d44]
  0x55f51c11df48: /usr/sbin/varnishd(+0x77f48) [0x55f51c11df48]
  0x55f51c11e458: /usr/sbin/varnishd(+0x78458) [0x55f51c11e458]
  0x7fde0138814a: /lib64/libpthread.so.0(+0x814a) [0x7fde0138814a]
  0x7fde010b9f23: /lib64/libc.so.6(clone+0x43) [0x7fde010b9f23]
pthread.self = 0x7fde0259b700
pthread.name = (cache-worker)
pthread.attr = {
  guard = 4096,
  stack_bottom = 0x7fde0258e000,
  stack_top = 0x7fde0259c000,
  stack_size = 57344,
}
thr.req = (nil) {
},
thr.busyobj = 0x7fdd323daea0 {
  end = 0x7fdd323eae80,
  retries = 0,
  sp = 0x7fddd1e87520 {
    fd = -1, vxid = 66196080,
    t_open = 1620814957.649612,
    t_idle = 1620814957.649801,
    ws = 0x7fddd1e87560 {
      id = \"ses\",
      {s, f, r, e} = {0x7fddd1e875b8, +96, (nil), +576},
    },
    transport = HTTP/1 {
      state = HTTP1::NewReq
    }
    client = 127.0.0.1 25054 127.0.0.1:6081,
  },
  worker = 0x7fde02599570 {
    ws = 0x7fde02599618 {
      id = \"wrk\",
      {s, f, r, e} = {0x7fde02598b60, +0, (nil), +2040},
    },
    VCL::method = BACKEND_RESPONSE,
    VCL::return = deliver,
    VCL::methods = {BACKEND_FETCH, BACKEND_RESPONSE},
  },
  vfc = 0x7fdd323dcda8 {
    failed = 0,
    req = 0x7fdd323db4b8,
    resp = 0x7fdd323db930,
    wrk = 0x7fde02599570,
    oc = 0x7fddd19f6a00,
    obj_flags = 0x0,
  },
  ws = 0x7fdd323daee0 {
    id = \"bo\",
    {s, f, r, e} = {0x7fdd323dcdf0, +5280, (nil), +57480},
  },
  ws_bo = 0x7fdd323dd970,
  http[bereq] = 0x7fdd323db4b8 {
    ws = 0x7fdd323daee0 {
      [Already dumped, see above]
    },
    hdrs {
    ```
 Full log
[panic1.log](https://github.com/varnishcache/varnish-cache/files/6478766/panic1.log)
[panic2.log](https://github.com/varnishcache/varnish-cache/files/6478767/panic2.log)

Do not hesitate to ask me for more information if needed.

**Comment by Dridi** [comment]
Looks like #3558, not part of any release.

**Comment by Dridi** [comment]
Bonjour Bastien, I checked the panic against the test case from #3558 and can confirm that it is indeed the same problem. This can happen with a slow backend if a revalidation is evaluated while the original fetch was still ongoing and failed (it doesn't have to be slow VCL like in the test case). One thing that may slow down a backend could be a heavy load, which is of course unfortunate. This bug fix will be part of the upcoming 6.0.8 release (see #3582) and unless you use a specific feature from the 6.5 series I would recommend to switch to the 6.0 LTS branch.
